README.md
[![build-status]](https://travis-ci.org/abhinav/farmhash)

This package provides bindings to the `farmhash` library.

FarmHash is a family of very fast hash functions (see [`farmhash`][]). They are
not suitable for cryptography.

The documentation for this package is available on [Hackage][] and [here][].

  [build-status]: https://travis-ci.org/abhinav/farmhash.svg?branch=master
  [`farmhash`]: https://code.google.com/p/farmhash/
  [Hackage]: http://hackage.haskell.org/package/farmhash
  [here]: http://abhinavg.net/farmhash/


CHANGES.md
0.1.0.5
=======

-   Support GHC 8.

0.1.0.4
=======

-   Loosen more version constraints.

0.1.0.3
=======

-   Loosen version constraints for `base`.

0.1.0.2
=======

-   Add `-fno-exceptions` while compiling C code. This prevents dynamic linking
    issues in the C++ code. (#1)

0.1.0.1
=======

-   Initial release.


test/FarmHashSpec.hs
module FarmHashSpec (spec) where

import Test.Hspec
import Control.Applicative
import Data.ByteString (ByteString)
import Test.QuickCheck
import Control.Monad
import Test.Hspec.QuickCheck

import qualified Data.ByteString as B

import FarmHash

byteStringOfLength :: NonNegative Int -> IO ByteString
byteStringOfLength (NonNegative l) = generate $ B.pack <$> vector l

spec :: Spec
spec = do
    checkBasic [
        ("hash", hash)
      , ("hash32", hash32)
      , ("fingerprint32", fingerprint32)
      ] 
    checkBasic [("hash64", hash64), ("fingerprint64", fingerprint64)]
    checkSeeded "hash32WithSeed" hash32WithSeed
    checkSeeded "hash64WithSeed" hash64WithSeed

    describe "hash64WithSeeds" $ do
        prop "operates on bytestrings" $ \len seed1 seed2 -> do
            s <- byteStringOfLength len
            hash64WithSeeds s seed1 seed2 `shouldSatisfy` (> 0)

        prop "operates on subsrings" $ \len seed1 seed2 tak drp -> do
            s <- byteStringOfLength len
            hash64WithSeeds (B.take tak (B.drop drp s)) seed1 seed2
                `shouldSatisfy` (> 0)
  where
    checkBasic h = forM_ h $ \(name, func) -> describe name $ do
        prop "operates on bytestrings" $ \len -> do
            s <- byteStringOfLength len
            func s `shouldSatisfy` (> 0)

        prop "operates on substrings" $ \tak drp len -> do
            s <- byteStringOfLength len
            func (B.take tak (B.drop drp s))
                `shouldSatisfy` (> 0)

    checkSeeded name func = describe name $ do
        prop "operates on bytestrings" $ \len seed -> do
            s <- byteStringOfLength len
            func s seed `shouldSatisfy` (> 0)

        prop "operates on subsrings" $ \len seed tak drp -> do
            s <- byteStringOfLength len
            func (B.take tak (B.drop drp s)) seed
                `shouldSatisfy` (> 0)


src/FarmHash.hs
-- |
-- Module      :  FarmHash
-- Copyright   :  (c) Abhinav Gupta 2015
-- License     :  BSD3
--
-- Maintainer  :  Abhinav Gupta <mail@abhinavg.net>
-- Stability   :  experimental
--
-- FarmHash is a family of very fast hash functions. They are not suitable for
-- cryptography.
--
-- See <https://code.google.com/p/farmhash/>.
module FarmHash (

    -- * Hash

    -- | These functions are platform-__dependent__. Their output may be
    -- different on different platforms for the same string.

      hash
    , hash32
    , hash32WithSeed

    , hash64
    , hash64WithSeed
    , hash64WithSeeds

    -- * Fingerprint

    -- | These functions are platform-__independent__. They are guaranteed to
    -- produce the same output for the same string on different platforms.

    , fingerprint32
    , fingerprint64
    ) where

import Data.ByteString (ByteString)
import Data.Word       (Word32, Word64)

import FarmHash.Internal

-- | Hash function for a byte array.
--
-- May change from time to time, and may differ on platforms.
hash :: ByteString -> Word32
hash = withByteString cHash

-- | Hash function for a byte array. Most useful in 32-bit binaries.
--
-- May change from time to time, and may differ on platforms.
hash32 :: ByteString -> Word32
hash32 = withByteString cHash32

-- | Hash function for a byte array. For convenience, a 32-bit seed is also
-- hashed into the result.
--
-- May change from time to time, and may differ on platforms.
hash32WithSeed :: ByteString -> Word32 -> Word32
hash32WithSeed = withByteString cHash32WithSeed

-- | Hash function for a byte array.
--
-- May change from time to time, and may differ on platforms.
hash64 :: ByteString -> Word64
hash64 = withByteString cHash64

-- | Hash function for a byte array. For convenience, a 64-bit seed is also
-- hashed into the result.
--
-- May change from time to time, and may differ on platforms.
hash64WithSeed :: ByteString -> Word64 -> Word64
hash64WithSeed = withByteString cHash64WithSeed

-- | Hash function for a byte array. For convenience, two seeds are also
-- hashed into the result.
--
-- May change from time to time, and may differ on platforms.
hash64WithSeeds :: ByteString -> Word64 -> Word64 -> Word64
hash64WithSeeds = withByteString cHash64WithSeeds

-- | Fingerprint function for a byte array. Most useful in 32-bit binaries.
--
-- Portable, forever-fixed hash function.
fingerprint32 :: ByteString -> Word32
fingerprint32 = withByteString cFingerprint32

-- | Fingerprint function for a byte array.
--
-- Portable, forever-fixed hash function.
fingerprint64 :: ByteString -> Word64
fingerprint64 = withByteString cFingerprint64


cbits/farmhash_c.cc
#define FARMHASH_NO_CXX_STRING
#include "farmhash.h"
#include "farmhash_c.h"

extern "C"
{

size_t Hash(const char* s, size_t len)
{
    return util::Hash(s, len);
}

uint32_t Hash32(const char* s, size_t len)
{
    return util::Hash32(s, len);
}

uint32_t Hash32WithSeed(const char* s, size_t len, uint32_t seed)
{
    return util::Hash32WithSeed(s, len, seed);
}

uint64_t Hash64(const char* s, size_t len)
{
    return util::Hash64(s, len);
}

uint64_t Hash64WithSeed(const char* s, size_t len, uint64_t seed)
{
    return util::Hash64WithSeed(s, len, seed);
}

uint64_t Hash64WithSeeds(const char* s, size_t len, uint64_t seed0, uint64_t seed1)
{
    return util::Hash64WithSeeds(s, len, seed0, seed1);
}

uint32_t Fingerprint32(const char* s, size_t len)
{
    return util::Fingerprint32(s, len);
}

uint64_t Fingerprint64(const char* s, size_t len)
{
    return util::Fingerprint64(s, len);
}

} /* end extern "C" */


src/FarmHash/Internal.hs
{-# LANGUAGE ForeignFunctionInterface #-}
module FarmHash.Internal
    ( cHash
    , cHash32
    , cHash32WithSeed

    , cHash64
    , cHash64WithSeed
    , cHash64WithSeeds

    , cFingerprint32
    , cFingerprint64

    , withByteString
    ) where

import Data.ByteString    (ByteString)
import Data.Word          (Word32, Word64, Word8)
import Foreign.ForeignPtr (withForeignPtr)
import Foreign.Ptr        (Ptr, plusPtr)
import System.IO.Unsafe   (unsafePerformIO)

import qualified Data.ByteString.Internal as BI


foreign import ccall unsafe "farmhash_c.h Hash"
    cHash :: Ptr Word8 -> Int -> Word32

foreign import ccall unsafe "farmhash_c.h Hash32"
    cHash32 :: Ptr Word8 -> Int -> Word32

foreign import ccall unsafe "farmhash_c.h Hash32WithSeed"
    cHash32WithSeed :: Ptr Word8 -> Int -> Word32 -> Word32

foreign import ccall unsafe "farmhash_c.h Hash64"
    cHash64 :: Ptr Word8 -> Int -> Word64

foreign import ccall unsafe "farmhash_c.h Hash64WithSeed"
    cHash64WithSeed :: Ptr Word8 -> Int -> Word64 -> Word64

foreign import ccall unsafe "farmhash_c.h Hash64WithSeeds"
    cHash64WithSeeds :: Ptr Word8 -> Int -> Word64 -> Word64 -> Word64

foreign import ccall unsafe "farmhash_c.h Fingerprint32"
    cFingerprint32 :: Ptr Word8 -> Int -> Word32

foreign import ccall unsafe "farmhash_c.h Fingerprint64"
    cFingerprint64 :: Ptr Word8 -> Int -> Word64

withByteString :: (Ptr Word8 -> Int -> b) -> ByteString -> b
withByteString f (BI.PS fp off len) =
    unsafePerformIO . withForeignPtr fp $ \p ->
        return $! f (p `plusPtr` off) len


scripts/update-gh-pages.sh
#!/bin/bash

set -e

findUp() {
	if [[ -f "$1/$2" ]]; then
		echo "$1"
	else
		findUp "$1/.." "$2"
	fi
}

ROOT=$(findUp "." "farmhash.cabal")

BASE="https://github.com/abhinav/farmhash/tree/master"

OPTIONS="\
	--source-base=\"$BASE/\" \
	--source-module=\"$BASE/%{FILE}\" \
	--source-entity=\"$BASE/%{FILE}#L%{LINE}\""

pushd "$ROOT/gh-pages"
git rm -rf .
popd

GHC_BIN="$(dirname "$(stack exec which ghc)")"

pushd "$ROOT"
PATH="$GHC_BIN:$PATH" runhaskell Setup.hs configure \
	--package-db="$(stack path --local-pkg-db)" \
	--package-db="$(stack path --snapshot-pkg-db)"

PATH="$GHC_BIN:$PATH" runhaskell Setup.hs haddock \
	--haddock-options="--odir=gh-pages $OPTIONS" \
	--html-location="http://hackage.haskell.org/packages/archive/\$pkg/latest/doc/html"
popd

pushd "$ROOT/gh-pages"
git add .
git commit -m "Documentation update at $(date '+%FT%T%z')"
popd

echo "Now run: rm -rf '$ROOT/dist'"


bench/bench-farmhash.hs
{-# LANGUAGE OverloadedStrings #-}
module Main (main) where

import Criterion.Main
import Data.Word      (Word32, Word64)

import qualified Data.ByteString as B

import FarmHash

main :: IO ()
main = defaultMain $ do
    -- Try strings of length 8 to 4096 in exponents of 8.
    stringLength <- take 4 . drop 1 $ iterate (*16) 1

    let string = B.replicate stringLength 120
        seed32 = fromIntegral stringLength :: Word32
        seed64 = fromIntegral stringLength :: Word64
    string `seq` seed32 `seq` seed64 `seq` return ()

    return $ bgroup ("length:" ++ show stringLength) [
        bgroup "32-bit" [
            bench "hash" $ whnf hash string
          , bench "hash32" $ whnf hash32 string
          , bench "hash32WithSeed" $ whnf (hash32WithSeed string) seed32
          , bench "fingerprint32" $ whnf fingerprint32 string
          ]
      , bgroup "64-bit" [
            bench "hash64" $ whnf hash64 string
          , bench "hash64WithSeed" $ whnf (hash64WithSeed string) seed64
          , bench "hash64WithSeeds" $
              whnf (hash64WithSeeds string seed64) seed64
          , bench "fingerprint64" $ whnf fingerprint64 string
          ]
      ]
